tests/neuro_san/client/data_driven_agent_test.py

```diff
@@ -0,0 +1,159 @@
+
+from tests.neuro_san.client.agent_evaluator import AgentEvaluator
+from tests.neuro_san.client.agent_evaluator_factory import AgentEvaluatorFactory
+from tests.neuro_san.client.assert_forwarder import AssertForwarder
+
+
+class DataDrivenAgentTest:
+    """
+    Abstract test class whose subclasses define hocon file to parse as test cases.
+    """
+
+    TEST_KEYS: List[str] = ["text", "sly_data"]
+
+    def __init__(self, asserts: AssertForwarder, fixtures: FileOfClass = None):
+        """
+        Constructor
+        :param asserts: The AssertForwarder instance to use to integrate failures
+                        back into the test system.
+        :param fixtures: Optional path to the fixtures root.
+        """
+        self.asserts: AssertForwarder = asserts
+        self.fixtures: FileOfClass = fixtures
+        if self.fixtures is None:
+            self.fixtures = FileOfClass(__file__, path_to_basis="../../fixtures")
+
+    def one_test(self, hocon_file: str):
+        """
+        Use a single hocon file in the fixtures as a test case"
+
+        :param hocon_file: The name of the hocon from the fixtures directory.
+        """
+        test_case: Dict[str, Any] = self.parse_hocon_test_case(hocon_file)
+
+        # Get the agent to use
+        agent: str = test_case.get("agent")
+        self.asserts.assertIsNotNone(agent)
+
+        # Get the connection type
+        connections: Union[List[str], str] = test_case.get("connections")
+        if connections is None:
+            # Assume direct if not specified
+            connections = ["direct"]
+        elif isinstance(connections, str):
+            # Make single strings into a list for consistent parsing
+            connections = [connections]
+        self.asserts.assertIsInstance(connections, List)
+        self.asserts.assertTrue(len(connections) > 0)
+
+        # Collect the interations to test for
+        empty: List[Any] = []
+        interactions: List[Dict[str, Any]] = test_case.get("interactions", empty)
+        self.asserts.assertTrue(len(interactions) > 0)
+
+        # Collect other session information
+        use_direct: bool = test_case.get("use_direct", False)
+        metadata: Dict[str, Any] = test_case.get("metadata", None)
+
+        for connection in connections:
+
+            session: AgentSession = AgentSessionFactory().create_session(connection,
+                                                                         agent,
+                                                                         use_direct=use_direct,
+                                                                         metadata=metadata)
+            chat_context: Dict[str, Any] = None
+            for interaction in interactions:
+                if isinstance(session, DirectAgentSession):
+                    session.reset()
+                chat_context = self.interact(agent, session, interaction, chat_context)
```

Comment:
It's quite likely these nested loops will get broken up in the future. For now it allows for @vince-leaf 's requirements of testing different connection mechanisms with the same test.